File: REPASO ESPECIAL/Python/Ejemplo_1/resultado.py
# --- CLASES BASE (Proporcionadas por el examen) ---
class Song:
    """Representa una canción con título, artista y duración en segundos."""
    def __init__(self, title, artist, duration):
        self.title = title
        self.artist = artist
        self.duration = duration

    def __str__(self):
        return f"{self.title} by {self.artist} ({self.duration}s)"

class Node:
    """Un nodo en la lista enlazada, contiene una canción."""
    def __init__(self, song):
        self.value = song
        self.next = None

class PlaylistBase:
    """Clase base para la gestión de la lista enlazada de canciones."""
    def __init__(self):
        self.first = None
        self._size = 0

    def add_song(self, song):
        new_node = Node(song)
        if self.first is None:
            self.first = new_node
        else:
            current = self.first
            while current.next:
                current = current.next
            current.next = new_node
        self._size += 1

    def get_song(self, index):
        if index < 0 or index >= self._size:
            return None
        current = self.first
        for _ in range(index):
            current = current.next
        return current.value

    def get_size(self):
        return self._size

    def __str__(self):
        if self.first is None:
            return "Playlist vacía"
        items = []
        current = self.first
        while current:
            items.append(str(current.value))
            current = current.next
        return " -> ".join(items)

# --- CLASE A DESARROLLAR (Nuestra Solución) ---
class Playlist(PlaylistBase):
    """Gestiona una playlist con funcionalidades avanzadas."""
    
    # Atributo de CLASE: compartido por todas las instancias de Playlist
    _total_played_time = 0

    def __init__(self):
        super().__init__()
        # Atributo de INSTANCIA: único para cada objeto Playlist
        self._time_spent = 0

    @property
    def total_duration(self):
        """(Solo lectura) Calcula la duración total de las canciones restantes."""
        total = 0
        current_node = self.first
        while current_node:
            total += current_node.value.duration
            current_node = current_node.next
        return total

    def play_song(self, index):
        """Elimina una canción, la devuelve y actualiza los contadores de tiempo."""
        if index < 0 or index >= self.get_size():
            return None

        song_to_play = None

        if index == 0:
            song_to_play = self.first.value
            self.first = self.first.next
        else:
            previous_node = None
            current_node = self.first
            count = 0
            while count < index:
                previous_node = current_node
                current_node = current_node.next
                count += 1
            song_to_play = current_node.value
            previous_node.next = current_node.next
        
        self._size -= 1 # No olvidar decrementar el tamaño

        if song_to_play:
            self._time_spent += song_to_play.duration
            Playlist._total_played_time += song_to_play.duration
        
        return song_to_play

    @classmethod
    def get_total_played_time(cls):
        """Devuelve el tiempo total reproducido en TODAS las playlists."""
        return cls._total_played_time

    def get_playlist_as_list(self):
        """Devuelve una lista de Python con las canciones de la playlist."""
        python_list = []
        current_node = self.first
        while current_node:
            python_list.append(current_node.value)
            current_node = current_node.next
        return python_list

File: REPASO ESPECIAL/Java/Ejemplo_1/resultado.java
/**
 * La clase Inventory gestiona el inventario de productos de una tienda.
 * Hereda de InventoryBase para reutilizar la estructura de la lista enlazada.
 * Añade funcionalidades para vender, reponer, y analizar el inventario.
 */
public class Inventory extends InventoryBase {

    // --- ATRIBUTOS ---

    /**
     * Almacena los ingresos totales generados por las ventas en ESTA instancia
     * de inventario. Es una variable de instancia, cada objeto Inventory
     * tendrá su propia copia.
     */
    private double totalRevenue = 0.0;

    /**
     * Almacena el número total de unidades de producto vendidas en TODAS
     * las instancias de Inventory. Es 'static', lo que significa que es una
     * variable compartida por todos los objetos de esta clase.
     */
    private static int totalProductsSold = 0;


    // --- MÉTODOS IMPLEMENTADOS ---

    /**
     * Vende una cantidad específica de un producto, buscándolo por su ID.
     *
     * @param productId El ID del producto a vender.
     * @param quantity  La cantidad de unidades a vender.
     * @return El valor total de la venta (precio * cantidad) si fue exitosa,
     * o 0.0 si el producto no se encontró o no había stock suficiente.
     */
    public double sellProduct(int productId, int quantity) {
        // Empezamos el recorrido desde el primer nodo de la lista.
        Node current = first;

        // Recorremos la lista mientras no lleguemos al final (current != null).
        while (current != null) {
            Product product = current.value; // Obtenemos el producto del nodo actual.

            // Comprobamos si el ID del producto actual coincide con el que buscamos.
            if (product.productId == productId) {
                // ¡Producto encontrado! Ahora verificamos si hay stock suficiente.
                if (product.quantity >= quantity) {
                    // Hay stock. Procedemos con la venta.
                    product.quantity -= quantity; // 1. Reducimos el stock del producto.

                    double saleValue = product.price * quantity; // 2. Calculamos el valor de la venta.

                    this.totalRevenue += saleValue; // 3. Sumamos al ingreso de esta tienda.
                    Inventory.totalProductsSold += quantity; // 4. Sumamos al contador global de ventas.

                    return saleValue; // 5. Devolvemos el valor de la venta y terminamos.
                } else {
                    // No hay suficiente stock, la venta no es posible.
                    return 0.0;
                }
            }
            // Si no es el producto que buscamos, avanzamos al siguiente nodo.
            current = current.next;
        }

        // Si el bucle termina, significa que el producto no estaba en la lista.
        return 0.0;
    }

    /**
     * Calcula el valor monetario total de todo el inventario.
     * (Suma de la cantidad * precio de cada producto).
     *
     * @return El valor total del inventario como un double.
     */
    public double getTotalInventoryValue() {
        double totalValue = 0.0;
        Node current = first; // Empezamos el recorrido.

        while (current != null) {
            Product product = current.value;
            // Sumamos el valor del stock de este producto al total.
            totalValue += product.quantity * product.price;
            current = current.next; // Avanzamos al siguiente.
        }
        return totalValue;
    }

    /**
     * Aumenta el stock de un producto existente, buscándolo por su ID.
     *
     * @param productId El ID del producto a reponer.
     * @param quantity  La cantidad de unidades a añadir al stock.
     * @return true si el producto fue encontrado y actualizado, false en caso contrario.
     */
    public boolean restockProduct(int productId, int quantity) {
        Node current = first; // Empezamos el recorrido.

        while (current != null) {
            Product product = current.value;
            if (product.productId == productId) {
                // ¡Producto encontrado!
                product.quantity += quantity; // Aumentamos su stock.
                return true; // Indicamos que la operación fue exitosa.
            }
            current = current.next;
        }

        // Si el bucle termina, el producto no se encontró.
        return false;
    }

    /**
     * Devuelve un array con todos los productos cuyo stock está por debajo de un umbral.
     *
     * @param threshold El límite de stock.
     * @return Un array de objetos Product que cumplen la condición.
     */
    public Product[] getProductsBelowStock(int threshold) {
        // --- Primer recorrido: Contar para saber el tamaño del array ---
        int count = 0;
        Node current = first;
        while (current != null) {
            if (current.value.quantity < threshold) {
                count++;
            }
            current = current.next;
        }

        // --- Segundo recorrido: Crear el array y rellenarlo ---
        Product[] lowStockProducts = new Product[count]; // Creamos el array del tamaño exacto.
        int index = 0;
        current = first; // Reiniciamos el recorrido.
        while (current != null) {
            if (current.value.quantity < threshold) {
                lowStockProducts[index] = current.value; // Añadimos el producto al array.
                index++; // Preparamos el índice para el siguiente.
            }
            current = current.next;
        }

        return lowStockProducts;
    }
}
